### PR TITLE
Create a trait that grants a condition on actor creation

### PR DESCRIPTION
**Pull request by penev92**
This grants the actor a random condition for a list in order to enable having different "variants" of an actor (like the Technicals in Generals).

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnCreate.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        readonly GrantConditionOnCreateInfo info;
+
+        public GrantConditionOnCreate(Actor self, GrantConditionOnCreateInfo info)
+        {
+            this.info = info;
+        }
+
+        public void Created(Actor self)
+        {
+            if (!info.Conditions.Any())
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnCreate.cs]: Does this even make sense if we have `[FieldLoader.Require]` on that field?

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnCreate.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                return;
+
+			var condition = info.Conditions.Random(self.World.SharedRandom);
+			var conditionManager = self.TraitOrDefault<ConditionManager>();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnCreate.cs]: Either use `.Trait` or add a `null` check for `conditionManager`.

**Review by abcdefg30 — DISMISSED** [review]
Please change the spaces to tabs.

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnCreate.cs (hunk 1 above)
Comment (penev92) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnCreate.cs]: No, I tested that. If you specify in the YAML
`Conditions:` and leave it blank the `Required` check passes, but there is still no value.

**Comment by MustaphaTR** [comment]
I think GrantRandomConditionOnCreation or at least GrantConditionOnCreation sounds better as a name for trait.

**Comment by GraionDilach** [comment]
This is absolutely NOT modder friendly however. I'm aware @pchote propagated this for a good while now, but this is just another tunnel vision reimplementation with limited modding options compared to the original, akin to how the TS EMP Cannon turned out.

Sure, conditions might be the future and all that, but due to the sheer amount of visual offsets defined in yaml trait codes (turret offsets, firing offsets), actor variations cannot be done effectively via random conditions - this limits their use and requires really finicky yaml code to reach the level of complexity Generals could do.

Take note that Generals could also spawn entirely different units as variations - a good example is the Rise of the Reds Propaganda Airship's Trollship variant, which won't even get selected with the "select all of this unit's type" button, because it isn't defined as a mere variation.

**Comment by GraionDilach** [comment]
Footnote: his was the feature which pushed me to start my soft-fork - as I said back at https://github.com/OpenRA/OpenRA/issues/12554#issuecomment-271105904 - , because I realized https://github.com/AttacqueSuperior/Engine/pull/18 wouldn't be "compliant" and favorable over this solution even though it would actually be able to do everything Generals could and allow YAML deduplication via creating an unit template for inheritance and individual variants would only need to define offsets/variant codes.

**Comment by pchote** [comment]
> OpenRA.Utility(1,1): Error: Missing Type: GrantConditionOnCreateInfo

**Comment by GraionDilach** [comment]
How does a mapper preplace a specific unit variant in this system?

**Comment by penev92** [comment]
@GraionDilach ok, so this approach is somewhat harder to use for really complicated cases where you want really different units (and simpler to use for simple cases). Does that mean it's bad to have, though?

**Comment by GraionDilach** [comment]
My real issue with it is that it works "like in Generals" yet it can't be used to spawn completely different actors or allow mappers to choose a variant statically either - which Generals could do. These are just arbitrary limits introduced for no reason - and actually hurts usability more than a default magic value on a trait.

I don't believe it is simpler to use for simple cases either - it is simpler to use if you don't rely on trait conditions a lot and don't have as complicated gameplay mechanics as Generals had (I might be doing it wrong, but I am already at the level where I'm creating shim conditions to simplify my trait setups, because no way I'd write rank-level || empdisable || weaponjammer || temporal || paradrop for all weapons defined and just shim emp-temporal-weaponjammers-paradrop to a weapondisable onto a template).

Feel free taking that other option suggested on the IRC anyway. That's a clearly working solution to the issue I present. It's working for good months now afterall.

**Comment by pchote** [comment]
> I'm creating shim conditions to simplify my trait setups, because no way I'd write rank-level || empdisable || weaponjammer || temporal || paradrop for all weapons defined and just shim emp-temporal-weaponjammers-paradrop to a weapondisable onto a template

This was how I intended people to do things, so :+1:

**Comment by Arular101** [comment]
Please don't forget to update the copyright notice year.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f4c840a15bede8652e53fb34a69004d6d04d8527...ec36fab85503620482d21519f83856a249663d3a

**Pull request merged**